pkg/apiserver/storage/ram/store.go

```diff
@@ -266,10 +275,46 @@ func (s *store) dispatchEvents() {
 }
 
 func (s *store) dispatchEvent(event antreastorage.InternalEvent) {
-	s.watcherMutex.Lock()
-	defer s.watcherMutex.Unlock()
-	// TODO: Optimize this to dispatch the event based on watchers' selector.
-	for _, watcher := range s.watchers {
-		watcher.add(event)
+	var failedWatchers []*storeWatcher
+
+	func() {
+		s.watcherMutex.RLock()
+		defer s.watcherMutex.RUnlock()
+
+		// First try to send events without blocking, to avoid setting up a timer
+		// for every event.
+		// blockedWatchers keeps watchers whose buffer are full.
+		var blockedWatchers []*storeWatcher
+		// TODO: Optimize this to dispatch the event based on watchers' selector.
+		for _, watcher := range s.watchers {
+			if !watcher.nonBlockingAdd(event) {
+				blockedWatchers = append(blockedWatchers, watcher)
+			}
+		}
+		if len(blockedWatchers) == 0 {
+			return
+		}
+		klog.V(2).Infof("%d watchers were not available to receive event %+v immediately", len(blockedWatchers), event)
+
+		// Then try to send events to blocked watchers with a timeout. If it
+		// timeouts, it means the watcher is too slow to consume the events or the
+		// underlying connection is already dead, terminate the watcher in this case.
+		timer := time.NewTimer(watcherAddTimeout)
```

Comment:
isn't `context.WithTimeout` a more idiomatic way of implementing timeouts in Go?

Reply:
I think you are right, the [implementation ](https://github.com/kubernetes/kubernetes/blob/6099a739d8cb1762cdadf956fae484833b937c5d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L899)I referred to must use Timer because it wants to reuse it across events to avoid re-allocation, but I made it a local variable then it doesn't make sense to use Timer any more as it only fires once and I need to set it to nil manually to indicate it has fired while context.Done() can be called infinite number of times.
Now I think I should follow the implementation completely to leverage Timer's advantage: it can be reset and reused when it's used frequently.